### PR TITLE
Support saving input_example with no conversion

### PR DESCRIPTION
**Pull request by serena-ruan**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/serena-ruan/mlflow/pull/10741?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/10741/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 10741
```

</p>
</details>

### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?
Due to backwards compatibility issue, we convert input_example into pandas DataFrame by default. This is to add a param that supports saving input_example directly without converting it.

### How is this PR tested?

- [ ] Existing unit/integration tests
- [ ] New unit/integration tests
- [x] Manual tests

Tested on dogfood that it saves input_example directly.
image

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [x] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

image
image

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/langchain/__init__.py [hunk 1]
```diff
@@ -98,6 +98,7 @@ def save_model(
     metadata=None,
     loader_fn=None,
     persist_dir=None,
+    **kwargs,
```
Comment (harupy) [inline mlflow/langchain/__init__.py:101]:
```suggestion
    example_no_conversion=False,
```


Can we add `example_no_conversion` as a keyword argument?

File: mlflow/models/utils.py [hunk 2]
```diff
@@ -336,9 +338,25 @@ def _save_example(mlflow_model: Model, input_example: ModelInputExample, path: s
...
+            "type": "json_object",
+        }
+        try:
+            with open(os.path.join(path, example_info[INPUT_EXAMPLE_PATH]), "w") as f:
+                json.dump(input_example, f, cls=NumpyEncoder)
+        except Exception as e:
+            raise MlflowException.invalid_parameter_value(
+                "Failed to save input example. Please make sure the input example is jsonable "
+                f"when no_conversion is True. Got error: {e}"
+            )
```
Comment (BenWilson2) [inline mlflow/models/utils.py:353]: Should we preserve the original traceback here?

```
raise MlflowException.invalid_parameter_value(f"... Got error: {e}") from e
```

File: mlflow/langchain/__init__.py [hunk 3]
```diff
@@ -306,6 +307,7 @@ def log_model(
     metadata=None,
     loader_fn=None,
     persist_dir=None,
+    example_no_conversion=False,
```
Comment (BenWilson2) [inline mlflow/langchain/__init__.py:310]: Can we add this to the docstrings for both `log_model` and `save_model`?

File: mlflow/openai/__init__.py [hunk 4]
```diff
@@ -457,6 +458,7 @@ def log_model(
     pip_requirements=None,
     extra_pip_requirements=None,
     metadata=None,
+    example_no_conversion=False,
```
Comment (BenWilson2) [inline mlflow/openai/__init__.py:461]: docstring adds here as well

**Comment by BenWilson2** [comment]
Can we add a brief explanation of this behavior to https://www.mlflow.org/docs/latest/models.html#model-input-example as reference for users and show the difference between the two boolean states of the parameter with a simple example dict input (utilizing the screenshot in the PR content block above would be a good visual aid :) )

**Review by BenWilson2 — APPROVED** [review]
LGTM!

**Pull request merged**